Modify the return type of RunFilterPlugins to *Status

kerthcet opened the pull request:
Before, the return type of RunFilterPlugins is a Map, but considering we'll return immediately once we met unsuccessful status, this is not necessary.

Signed-off-by: Kante Yin <kerthcet@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature
/kind cleanup
/sig scheduling

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This is a follow-up of https://github.com/kubernetes/kubernetes/pull/114518#discussion_r1050162144
cc @ahg-g @Huang-Wei 

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

After this PR, we'll also have to make a adaption to cluster-autoscaler.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

kerthcet commented on pkg/scheduler/framework/runtime/framework_test.go:
```diff
@@ -1591,19 +1580,24 @@ func TestFilterPlugins(t *testing.T) {
 		},
 	}
 
+	cmpOpts := []cmp.Option{
```
We can't use `cmp.Diff(wantStatus, gotStatus)` directly because `status.Equal()` will compare the error via `errors.Is()`. 
```
// An error is considered to match a target if it is equal to that target or if
// it implements a method Is(error) bool such that Is(target) returns true.
```

We met this problem in several test cases, but I haven't thought of any better ideas yet. Some experiments like:
1) Compare the error message for simplify, AFAIK `status.Equal()` only used in tests.
2) Introduce a new error type embedded to status which implements `func (m MyError) Is(target error) bool`.

x13n commented:
The resulting change on autoscaler side will be trivial, so approving from that perspective.

/approve

kerthcet commented:
/retest

kerthcet commented:
> The resulting change on autoscaler side will be trivial, so approving from that perspective.
> 
> /approve

Thanks @x13n , I'll also follow this change in autoscaler after this got merged.

kerthcet commented on pkg/scheduler/framework/interface.go:
```diff
@@ -758,7 +758,7 @@ type PluginsRunner interface {
 	// removed from it to evaluate the possibility of preempting them to
 	// schedule the target pod.
-	RunFilterPlugins(context.Context, *CycleState, *v1.Pod, *NodeInfo) PluginToStatus
+	RunFilterPlugins(context.Context, *CycleState, *v1.Pod, *NodeInfo) *Status
```
We can't remove it yet, for `Unschedulable` and `UnschedulableAndUnresolvable` status, we still need the `Merge()` method.

ahg-g replied:
The type isn't use in any public interface, so we need to remove it. As for the status, that is easy to do inside RunPostFilterPlugins.

kerthcet replied:
Can we delay the removal of `PluginToStatus` until https://github.com/kubernetes/kubernetes/issues/113854, then as planned, we'll return directly when status is `UnschedulableAndUnresolvable`, no need of `Merge()` and `PluginToStatus` at that moment.

ahg-g replied:
ok, we can do this cleanup as part of #113854

ahg-g replied:
I just looked at https://github.com/kubernetes/kubernetes/issues/113854; what is the ask? the unschedulable statuses are already handled correctly (it returns `UnschedulableAndUnresolvable` if at least one PostFilter returns that status)

ahg-g replied:
ok, the ask is to return early, lets merge this then and we can discuss on https://github.com/kubernetes/kubernetes/pull/114699 the next steps.

kerthcet commented:
/kind feature

ahg-g commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -722,22 +722,21 @@ func (f *frameworkImpl) RunFilterPlugins(
 	nodeInfo *framework.NodeInfo,
-) framework.PluginToStatus {
-	statuses := make(framework.PluginToStatus)
+) (status *framework.Status) {
```
Named return values are recommended only for short functions (couple of lines)

kerthcet replied:
Updated.

ahg-g commented:
pls squash

kerthcet committed:
Modify the return type of RunFilterPlugins to *Status

Before, the return type of RunFilterPlugins is a Map, but considering we'll return immediately
once we met unsuccessful status, this is not necessary.

Signed-off-by: Kante Yin <kerthcet@gmail.com>

kerthcet commented:
I was sick last week. squashed @ahg-g

ahg-g commented:
/lgtm
/approve

Thanks! I hope you are doing better now.

ahg-g commented:
@kerthcet please open an issue in the CA repo to track the changes on their side.

Huang-Wei commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -722,22 +722,23 @@ func (f *frameworkImpl) RunFilterPlugins(
-) framework.PluginToStatus {
-	statuses := make(framework.PluginToStatus)
+) *framework.Status {
+	var status *framework.Status
```
stylish nit:

```go
	for _, pl := range f.filterPlugins {
		status := f.runFilterPlugin(ctx, pl, state, pod, nodeInfo)
		if !status.IsSuccess() {
			if !status.IsUnschedulable() {
				// Filter plugins are not supposed to return any status other than
				// Success or Unschedulable.
				status = framework.AsStatus(fmt.Errorf("running %q filter plugin: %w", pl.Name(), status.AsError()))
			}
			status.SetFailedPlugin(pl.Name())
			return status
		}
	}

	return nil
```

kerthcet commented:
> @kerthcet please open an issue in the CA repo to track the changes on their side.

Created https://github.com/kubernetes/autoscaler/issues/5397
